kubernetes/data_source_kubernetes_all_namespaces.go

```diff
@@ -0,0 +1,56 @@
+package kubernetes
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func dataSourceKubernetesAllNamespaces() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceKubernetesAllNamespacesRead,
+		Schema: map[string]*schema.Schema{
+			"namespaces": {
+				Type:        schema.TypeList,
+				Description: "List of all namespaces in a cluster.",
+				Computed:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+		},
+	}
+}
+
+func dataSourceKubernetesAllNamespacesRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*KubeClientsets).MainClientset
+
+	log.Printf("[INFO] Listing namespaces")
+	nsRaw, err := conn.CoreV1().Namespaces().List(metav1.ListOptions{})
+	if err != nil {
+		log.Printf("[DEBUG] Received error: %#v", err)
+		return err
+	}
+	namespaces := make([]string, len(nsRaw.Items))
+	for i, v := range nsRaw.Items {
+		namespaces[i] = string(v.Name)
+	}
+	log.Printf("[INFO] Received namespaces: %#v", namespaces)
+	err = d.Set("namespaces", namespaces)
+	if err != nil {
+		return err
+	}
+	idsum := sha256.New()
+	for _, v := range namespaces {
+		_, err := idsum.Write([]byte(v))
+		if err != nil {
+			return err
+		}
+	}
+	id := fmt.Sprintf("%x", idsum.Sum(nil))
+	d.SetId(id)
```

Comment:
No guidelines that I know of. I just went for this approach so that no changes are "registered" if the data (list of namespaces) didn't change.

I can see the motivations behind the Google (save an API call if the query filter didn't change) and the AWS one (ensure a fresh query is always done). I guess the AWS approach would have worked for this case as well.